File: traccia/aes/provider.py
"""
TRACCIA AES Adapter - AttributionProvider 接口与 AES 实现
"""

from abc import ABC, abstractmethod
from typing import Optional
from .models import ResponsibilityRecord


class AttributionProvider(ABC):
    """
    责任归属提供者（抽象接口）。

    Traccia 定义此接口，不依赖任何特定实现。
    AES 只是其中一种实现。
    """

    @abstractmethod
    def resolve_session(self, session_id: str) -> Optional[ResponsibilityRecord]:
        """解析 Session 的责任归属"""
        ...

    @abstractmethod
    def resolve_event(self, event_id: str) -> Optional[ResponsibilityRecord]:
        """解析 Event 的责任归属"""
        ...


class AESProvider(AttributionProvider):
    """
    AES 责任归属提供者（Stub 实现）。

    当前为存根实现，返回模拟数据。
    生产环境需接入真实 AES 服务端点。
    """

    def __init__(self, aes_endpoint: str = "http://localhost:8080"):
        self.aes_endpoint = aes_endpoint

    def resolve_session(self, session_id: str) -> Optional[ResponsibilityRecord]:
        """
        解析 Session 责任。

        存根实现：返回示例数据。
        生产实现：调用 AES API。
        """
        # Stub: 模拟 AES 响应
        from .models import Actor
        return ResponsibilityRecord(
            subject_id=session_id,
            subject_type="session",
            executor=Actor(id="agent_001", type="agent"),
            supervisor=Actor(id="operator_001", type="human"),
            approver=Actor(id="ceo_001", type="human"),
            policy_id="AES-RISK-001",
            risk_level="medium"
        )

    def resolve_event(self, event_id: str) -> Optional[ResponsibilityRecord]:
        """
        解析 Event 责任。

        存根实现：返回示例数据。
        生产实现：调用 AES API。
        """
        from .models import Actor
        return ResponsibilityRecord(
            subject_id=event_id,
            subject_type="event",
            executor=Actor(id="agent_001", type="agent"),
            risk_level="low"
        )

File: traccia-rust/crates/traccia-aes/src/provider.rs
use crate::models::{Actor, ResponsibilityRecord};

/// AttributionProvider：责任归属提供者（抽象 trait）
///
/// Traccia 定义此接口，不依赖任何特定实现。
/// AES 只是其中一种实现。
pub trait AttributionProvider {
    fn resolve_session(&self, session_id: &str) -> Option<ResponsibilityRecord>;
    fn resolve_event(&self, event_id: &str) -> Option<ResponsibilityRecord>;
}

/// AESProvider：AES 责任归属提供者（Stub 实现）
pub struct AESProvider {
    #[allow(dead_code)]
    endpoint: String,
}

impl AESProvider {
    pub fn new(endpoint: &str) -> Self {
        Self {
            endpoint: endpoint.to_string(),
        }
    }
}

impl AttributionProvider for AESProvider {
    fn resolve_session(&self, _session_id: &str) -> Option<ResponsibilityRecord> {
        // Stub: 模拟 AES 响应
        Some(
            ResponsibilityRecord::new(_session_id, "session", Actor::new("agent_001", "agent"))
                .with_supervisor(Actor::new("operator_001", "human"))
                .with_approver(Actor::new("ceo_001", "human"))
                .with_policy("AES-RISK-001")
                .with_risk("medium")
        )
    }

    fn resolve_event(&self, _event_id: &str) -> Option<ResponsibilityRecord> {
        // Stub: 模拟 AES 响应
        Some(
            ResponsibilityRecord::new(_event_id, "event", Actor::new("agent_001", "agent"))
                .with_risk("low")
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_aes_provider_resolve_session() {
        let provider = AESProvider::new("http://localhost:8080");
        let record = provider.resolve_session("sess-1").unwrap();

        assert_eq!(record.executor.id, "agent_001");
        assert!(record.supervisor.is_some());
        assert!(record.approver.is_some());
        assert_eq!(record.risk_level, "medium");
    }

    #[test]
    fn test_aes_provider_resolve_event() {
        let provider = AESProvider::new("http://localhost:8080");
        let record = provider.resolve_event("evt-1").unwrap();

        assert_eq!(record.subject_type, "event");
        assert_eq!(record.risk_level, "low");
    }
}
